### ISSUE TITLE
[Feature] Use JS to focus on input widgets

### ISSUE THREAD

**Issue by areyabhishek** [appsmithorg/appsmith#4874]
## Summary
Give users the ability to focus on an input widget. Something like text1.focus() would be useful.

**Comment by somangshu** [comment]
@areyabhishek are we talking about autofocus on input or trigger focus as an action?

**Comment by areyabhishek** [comment]
I'm talking about trigger focus as an action

**Comment by makulais** [comment]
I am finding this request in multiple places, are any of them being worked on?

**Comment by ame-appsmith** [comment]
This feature was requested by these users: [#1](https://app.intercom.com/a/apps/y10e7138/inbox/inbox/conversation/164629100163304#part_id=comment-164629100163304-16684073507) ; [#2](https://app.intercom.com/a/apps/y10e7138/inbox/inbox/conversation/164629100163321#part_id=comment-164629100163321-16939229476)

**Comment by gebnar** [comment]
I would also very much like to see this feature added.
I have run across multiple situations where I would like to automatically respond to user behavior by moving the cursor to an input box.
Specifically, I have deployed Appsmith pages in environments where staff are using barcode scanners to enter information. Sometimes I've been able to make it work. But if the user needs to manually select the correct text box between each scan, there is a lot of extra interaction required and our KPIs are significantly slower. Being able to auto-move the cursor to inputs would allow me to improve workflows by an order of magnitude in these circumstances.
Separately, shifting widget focus is just a generally useful option to have in UI development.

**Comment by nihaofana** [comment]
Seemed this focus request have been proposed long time ago. But why still not be realised? Is it too hard technically？ 
Or is there any other way to bypass it ,such as using a iframe to build a html input box?

**Comment by developer-honoru** [comment]
@bharath31
Waiting #17339 (comment) I wonder why it wasn't planned
do @rimildeyjsr for inventing new methods.
If there is no developer to do this, can I make one?
I need this feature

**Comment by bharath31** [comment]
@developer-honoru sorry for dropping the ball on this one. @satbir121 and @rimildeyjsr bumping this up ICYMI https://github.com/appsmithorg/appsmith/pull/17339#discussion_r992124173. Please help @developer-honoru with next steps.

**Comment by developer-honoru** [comment]
Thanks for checking! If there are not enough people, I want to make it myself. just give me guidelines @bharath31 @rimildeyjsr

**Comment by satbir121** [comment]
@developer-honoru This feature requires an architecture level change. Currently, we do not have any properties inside widgets that are mutable from JS. Everything is Read Only on Widgets. We are going to pick this up in first quarter of 2023. The underlying framework will be changed to make widget properties mutable.

[Set meta properties on widgets](https://github.com/appsmithorg/appsmith/issues/3532)

**Comment by Nikhil-Nandagopal** [comment]
@bharath31 another request for this which can be solved with #3532

**Comment by bharath31** [comment]
Another user asked for this on [discord](https://discord.com/channels/725602949748752515/778949891014656050/1110895423322857574)

**Comment by bharath31** [comment]
@Rishabh-Rathod @Druthi now that we've released widget property setters, should this be an easy solve?

**Comment by Rishabh-Rathod** [comment]
@bharath31 to support the above feature. 

1. We will need to internally introduce `isFocused` property for the list of widgets where this functionality needs to be added. 
2. Widget developers will need to make sure on the change of `isFocused` property the widget changes its focus state.
3. Add `setFocused` to the setter config. ( this task could be picked up by both the widget team or FE team )

cc: @somangshu

**Comment by somangshu** [comment]
@Rishabh-Rathod once the platform functionality is complete, the widgets team and use the API and decide on widgets that need the implementation. 

Let us know once the first bit is completed.

**Comment by Rishabh-Rathod** [comment]
@somangshu introducing `isFocused` ( new property ) to widget can also be taken care by widget developer as done before. 

@bharath31 Could you please mention list of all the widgets that need `setFocused` setter method.

**Comment by somangshu** [comment]
@Rishabh-Rathod I disagree. 
The requirement is to use JS to be able to focus on the widget. So this is a meta prop that should be allowed to set from the JS functions. The widgets team will not have the expertise to contribute on that front. If we have the API, widgets team can decide where all this needs to be implemented, we dont really need help there. 

cc @sbalaji1192 please correct me if I going wrong here.

**Comment by somangshu** [comment]
Also, `isFocused` prop is secondary here, that is an internal prop for users to check if a certain component / widget is focused. Whereas again, the requirement is to have an API to focus on the said component / widget programatically.

**Comment by Rishabh-Rathod** [comment]
@somangshu `isFocused` will be the property that would actually control the focus state of widget and would also reflect it. 

As soon as user start typing inside input widget, we will change the isFocused state to true and when the input widget is out of focus we will make the isFocused property as false. 

Now, `setFocused` will change the same isFocused state to control the focus state of widget. 
Hence, if the widget doesn’t support the `isFocused` state, we need to introduce it and make sure we change the value according to user action and vice versa i.e., on `isFocused` change we also change the widget focus. 

`setFocused` will be only responsible to update the isFocused state but the rest of the logic will need to be handled inside widget.

**Comment by sbalaji1192** [comment]
@somangshu I agree with Rishab, We need to make changes on the widget to consume the isFocused state.

**Comment by somangshu** [comment]
We need maybe another task that should be used to make the changes in widget.

**Comment by nirkons** [comment]
+1

**Comment by wigbertojr** [comment]
+1

**Comment by rohan-arthur** [comment]
@uliilu we haven't gotten to developing this feature yet. If setting the focus is crucial to your app building, you might also meanwhile want to try a [custom widget](https://docs.appsmith.com/build-apps/how-to-guides/Create-Custom-Widgets-Using-Iframe).

**Comment by uliilu** [comment]
@rohan-arthur, Hi, I will take a closer look to your suggestion. Thank you very much and all the best

**Comment by japrogramer** [comment]
@rohan-arthur I dont understand how custom widgets are suppose to solve this problem, 
I tried custom widget but the window object is not available there also,
the document object is also not available there either.

I've tried this a tag but the appsmith custom widget container refuses to connect at click, with an error

`<a t="#topContainer" className={"btn rounded-border"} onClick={handleNextPage} disabled={currentPage === totalPages}>Next</a>`


I've tried this but useRef is always null or compiled out,
```
    if (contentRef.current) {
      contentRef.current.scrollTo({
        top: 0,
        behavior: 'smooth'
      });
    }
```


**The only thing that comes close to working is This**
<font color="red">
### But this only works in the builder and not inside of the editor or in the deployed app
</font>

```
        const handleNextPage = (event) => {
	        scrollToTop(event)
          setCurrentPage((prevPage) => Math.min(prevPage + 1, totalPages));
        };
      
        const handlePrevPage = (event) => {
	        scrollToTop(event)
          setCurrentPage((prevPage) => Math.max(prevPage - 1, 1));
        };
  	const scrollToTop = (event) => {
		event.preventDefault();
		const button = event.target;
		const scrollableParent = findScrollableParent(button);
		if (scrollableParent) {
			scrollableParent.scrollTop = 0;
		}
	}

	const findScrollableParent = (element) => {
			if (!element) {
					return null;
			}

			// Check if the current element has the id 'contentTop'
			if (element.id === 'topContainer') {
					return element;
			}

			if (element.scrollHeight > element.clientHeight) {
					return element;
			} else {
					return findScrollableParent(element.parentElement);
			}
	}
```

**Comment by eedwardsSGS** [comment]
@mwoollen that seems like a good workaround if the scan workflow is always in the same sequence (which in many cases it can be).  We've got some extra complication as depending on the product some fields are disabled/enabled, but that's still manageable with some flags.

Thanks for the tip!  Still would love an official way of doing things, but this could improve what we're doing today in the meantime.

**Comment by hugorn** [comment]
+1

**Comment by hugorn** [comment]
I would like to click on the button and then clear the data and move the focus into that input to receive new data from the barcode reader.

<img width="444" height="706" alt="Image" src="https://github.com/user-attachments/assets/7c8c59c0-dea3-4ee2-afd8-9beb90ad5016" />

**Comment by MosheWelcher** [comment]
Any update? I would really like this feature to be implemented, especially because hitting tab sometimes moves it to weird places
